lib/User/UserEntry.php

```diff
@@ -0,0 +1,360 @@
+ * UserEntry is a wrapper around an ldap search array.
+ *
+ * It represents an LDAP user and is created by the UserManager on a getUsers
+ * call. It is more than a Plain Old PHP Objects (POPO) because it uses the ldap
+ * configuration to fetch the results fo different getters. It also does some
+ * sanity checking eg for dn, displayName, uuid, quota and home folder. Finally,
+ * it can store the internal ownCloud UID determined by the UserManager.
+ */
+class UserEntry {
+	/**
+	 * @var IConfig
+	 */
+	protected $config;
+	/**
+	 * @var ILogger
+	 */
+	protected $logger;
+	/**
+	 * @var Connection
+	 */
+	protected $connection;
+	/**
+	 * @var array
+	 */
+	protected $ldapEntry;
+	/**
+	 * @var string
+	 */
+	protected $ownCloudUID;
+
+	/**
+	 * @brief constructor, make sure the subclasses call this one!
+	 * @param IConfig $config
+	 * @param ILogger $logger
+	 * // FIXME Connection is used to look uf configuration ... pass in Configuration instead?
+	 * @param Connection $connection to lookup configured attribute names
+	 * @param array $ldapEntry an ldapEntry returned from Access::fetchListOfUsers()
+	 * @throws \InvalidArgumentException if entry does not contain a dn
+	 */
+	public function __construct(IConfig $config, ILogger $logger, Connection $connection, array $ldapEntry) {
+		$this->config = $config;
+		$this->logger = $logger;
+		$this->connection = $connection;
+		$this->ldapEntry = $ldapEntry;
+		// only accept an entry with the dn set
+		if ($this->getDN() === null) {
+			throw new \InvalidArgumentException('Entry must have the dn set');
+		}
+		//TODO extractAttributeValuesFromResult from Access
+	}
+
+	/**
+	 * @brief returns the Distinguished Name (DN) of the LDAP entry
+	 * @return string|null
+	 */
+	public function getDN() {
+		return $this->getAttributeValue ('dn', null);
+	}
+
+	/**
+	 * @see getOwncloudUID() to get the owncloud internal userid
+	 * @return string username as configured. THIS IS NOT the ownCloud internal userid!
+	 * @throws \OutOfBoundsException if username could not be determined
+	 */
+	public function getUsername() {
+		$attr = $this->getAttributeName('ldapExpertUsernameAttr');
+		if ($attr === '') {
+			$username = $this->getUUID(); // fallback to uuid
+		} else {
+			$username = $this->getAttributeValue($attr, null);
+		}
+		if ($username === null) {
+			$json = @json_encode($this->ldapEntry, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE | JSON_NUMERIC_CHECK | JSON_PARTIAL_OUTPUT_ON_ERROR );
+			if (json_last_error() !== JSON_ERROR_NONE) {
+				$json .= ', ' . json_last_error_msg();
+			};
+			throw new \OutOfBoundsException('Cannot determine username for '.$this->getDN().' from '.$json);
+		}
+		return $username;
+	}
+
+	/**
+	 * @brief returns the ownCloudUserID if it has been resolved by the Manager
+	 * @return string|null
+	 */
+	public function getOwnCloudUID() {
+		return $this->ownCloudUID;
+	}
+
+	/**
+	 * @param string $ownCloudUID
+	 */
+	public function setOwnCloudUID($ownCloudUID) {
```

Comment:
We need to store an already mapped uid somewhere.

Reply:
My point is if we can set the value only in the constructor in order to prevent modifications, or we need to modify the value once the object is already constructor.